manager/orchestrator/taskreaper/task_reaper.go

```diff
@@ -153,23 +178,53 @@ func (tr *TaskReaper) Run(ctx context.Context) {
 			}
 
 			if len(tr.dirty)+len(tr.cleanup) > maxDirty {
+				// stop the timer, so we don't fire it. if we get another event
+				// after we do this cleaning, we will reset the timer then
 				timer.Stop()
+				// if the timer had fired, drain out the value.
+				select {
+				case <-timer.C:
+				default:
+				}
+				isTimerStopped = true
 				tr.tick()
 			} else {
-				timer.Reset(reaperBatchingInterval)
+				if isTimerStopped {
```

Comment:
does it matter here if you always call Reset() without checking if the timer was stopped ?

Reply:
you've got the added and removed lines confused.

Reply:
Non-blocking - this code's logic seems fine, but looking at some of our vendored code like GRPC (see clientconn.go) it seems like the pattern is to create a timer at the beginning of the loop.  If the timer fires, set the timer to `nil`.  If the timer is not nil at the end of the loop, stop it and set it to `nil`, and there's no need to drain `timer.C`.  This does increase GC load, though, but it is easier to read and would be slightly less repeated code to drain the timer.

Reply:
@anshulpundir If I'm understanding the intent correctly, we want to clean up whenever `maxDirty` is reached or the timer ticks, whichever is sooner.  I think if we always called `Reset` without checking `isTimerStopped`, the the timer may never tick and we will just wait for `maxDirty` all the time (since calling `Reset` would mean that the the timer would be reset to start waiting another `reaperBatchingInterval` seconds every time an event comes in without hitting `maxDirty`.